zebra-state/src/service/check/anchors.rs

```diff
@@ -435,39 +454,67 @@ pub(crate) fn tx_anchors_refer_to_final_treestates(
     parent_chain: Option<&Arc<Chain>>,
     unmined_tx: &UnminedTx,
 ) -> Result<(), ValidateContextError> {
-    sapling_orchard_anchors_refer_to_final_treestates(
-        finalized_state,
-        parent_chain,
-        &unmined_tx.transaction,
-        unmined_tx.id.mined_id(),
-        None,
-        None,
-    )?;
+    if unmined_tx.transaction.has_sprout_joinsplit_data() {
+        let mut sprout_anchors_result = None;
+        let mut sapling_orchard_anchors_result = None;
+        rayon::in_place_scope_fifo(|s| {
+            // This check is expensive, because it updates a note commitment tree for each sprout JoinSplit.
+            // Since we could be processing attacker-controlled mempool transactions, we need to run each one
+            // in its own thread, separately from tokio's blocking I/O threads. And if we are under heavy load,
+            // we want verification to finish in order, so that later transactions can't delay earlier ones.
+            s.spawn_fifo(|_s| {
+                let mut sprout_final_treestates = HashMap::new();
+
+                fetch_sprout_final_treestates(
+                    &mut sprout_final_treestates,
+                    finalized_state,
+                    parent_chain,
+                    &unmined_tx.transaction,
+                    None,
+                    None,
+                );
```

Comment:
This is mainly a disk operation, so it should be on the tokio thread, not a rayon thread.

Reply:
Updated in [`b8d18d8`](https://github.com/ZcashFoundation/zebra/pull/5742/commits/b8d18d86450afd2d0ba15759017971a414f50553).